src/include/cache_inline.h

```diff
@@ -511,3 +511,70 @@ __wt_cache_eviction_check(WT_SESSION_IMPL *session, bool busy, bool readonly, bo
 
     return (__wt_cache_eviction_worker(session, busy, readonly, pct_full));
 }
+
+/*
+ * __wt_evict_busy_clr --
+ *     Clear the eviction busy field atomically. Restore the saved cursor if needed.
+ */
+static inline void
+__wt_evict_busy_clr(WT_SESSION_IMPL *session, WT_BTREE *btree, WT_CURSOR *save)
+{
+    (void)__wt_atomic_subv32(&btree->evict_busy, 1);
+    if (save != NULL)
+        /* Restore the caller's history store cursor. */
+        session->hs_cursor = save;
+}
+
+/*
+ * __wt_evict_busy_set --
+ *     Set the eviction busy field atomically. Before doing so make sure we have cached a history
+ *     store cursor. Save the session's original history store cursor if needed.
+ */
+static inline int
+__wt_evict_busy_set(WT_SESSION_IMPL *session, WT_BTREE *btree, WT_CURSOR **save)
+{
+    WT_CONNECTION_IMPL *conn;
+
+    conn = S2C(session);
+
+    /*
+     * If we have a history store cursor, save it. This ensures that if eviction needs to access the
+     * history store, it will get its own cursor, avoiding potential problems if it were to
+     * reposition or reset a history store cursor that we're in the middle of using for something
+     * else.
+     */
+    if (save != NULL) {
```

Comment:
The option to not save the history store cursor (i.e., if `save == NULL`) could cause problems because `wt_hs_cursor_open()`, asserts that `session->hs_cursor == NULL`.  

Three possible solutions, depending on how important you think it is to allow the `save == NULL` option.  

1. Don't allow `save == NULL` (None of the changes in this PR use that option.)
2. Add `session->hs_cursor == NULL` to the long precondition before the HS cursor open/close pair.
3. Refactor the `wt_hs_cursor_open()` and `wt_hs_cursor_close()` functions (they're quite simple) and create a `wt_hs_cursor_cache()` that just opens and closes a cursor on the HS btree to get it into the cursor cache, but doesn't touch `session->hs_cursor`.

Reply:
I have to backtrack some of these changes. They're again far too low and lead to deadlock. However, I think your suggestion about `wt_hs_cursor_cache()` is the way to go also. We need something that is not going to touch that. The other paths into `wt_evict` need it too and trying to manage the `session->hs_cursor` properly in all those places is going to get messy fast.